File: python/tests/helpers/mock_transport.py
from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator

from codex_client import JsonRpcCodec

_QUEUE_EOF = object()


class MockTransport:
    def __init__(self) -> None:
        self.sent_frames: list[str] = []
        self.closed = False
        self._queue: asyncio.Queue[object] = asyncio.Queue()

    async def send(self, data: str) -> None:
        self.sent_frames.append(data)

    def __aiter__(self) -> AsyncIterator[str]:
        return self._iterate()

    async def close(self) -> None:
        self.closed = True
        await self._queue.put(_QUEUE_EOF)

    async def inject(self, message: object) -> None:
        if isinstance(message, str):
            await self._queue.put(message)
            return
        await self._queue.put(JsonRpcCodec.encode(message))

    async def fail(self, exc: BaseException) -> None:
        await self._queue.put(exc)

    async def next_sent(self) -> dict[str, object]:
        while not self.sent_frames:
            await asyncio.sleep(0)
        return JsonRpcCodec.decode(self.sent_frames.pop(0))

    async def _iterate(self) -> AsyncIterator[str]:
        while True:
            item = await self._queue.get()
            if item is _QUEUE_EOF:
                return
            if isinstance(item, BaseException):
                raise item
            yield str(item)

File: typescript/tests/helpers/mock-transport.ts
import { JsonRpcCodec } from "../../src/codec.js";
import type { Transport } from "../../src/transport.js";

const QUEUE_EOF = Symbol("queue eof");

class AsyncQueue<T> {
  private readonly items: T[] = [];
  private readonly waiters: Array<(value: T) => void> = [];

  enqueue(value: T): void {
    const waiter = this.waiters.shift();
    if (waiter !== undefined) {
      waiter(value);
      return;
    }

    this.items.push(value);
  }

  dequeue(): Promise<T> {
    const value = this.items.shift();
    if (value !== undefined) {
      return Promise.resolve(value);
    }

    return new Promise((resolve) => {
      this.waiters.push(resolve);
    });
  }
}

export class MockTransport implements Transport {
  readonly sentFrames: string[] = [];
  closed = false;
  private readonly queue = new AsyncQueue<string | Error | typeof QUEUE_EOF>();

  async send(data: string): Promise<void> {
    this.sentFrames.push(data);
  }

  async close(): Promise<void> {
    this.closed = true;
    this.queue.enqueue(QUEUE_EOF);
  }

  async inject(message: unknown): Promise<void> {
    if (typeof message === "string") {
      this.queue.enqueue(message);
      return;
    }
    this.queue.enqueue(JsonRpcCodec.encode(message as never));
  }

  async fail(error: Error): Promise<void> {
    this.queue.enqueue(error);
  }

  async nextSent(): Promise<Record<string, unknown>> {
    while (this.sentFrames.length === 0) {
      await new Promise((resolve) => setTimeout(resolve, 0));
    }

    const frame = this.sentFrames.shift();
    if (frame === undefined) {
      throw new Error("Expected a sent frame");
    }
    return JsonRpcCodec.decode(frame) as unknown as Record<string, unknown>;
  }

  async *[Symbol.asyncIterator](): AsyncIterator<string> {
    while (true) {
      const item = await this.queue.dequeue();
      if (item === QUEUE_EOF) {
        return;
      }
      if (item instanceof Error) {
        throw item;
      }
      yield item;
    }
  }
}
